include/hip/spirv_hip_fp16.h

```diff
@@ -18,1011 +18,1713 @@ AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
-  __host__ __device__ __half& operator=(__half&&) = default;
-  __host__ __device__ __half& operator=(const __half_raw& x) {
-    data = x.data;
-    return *this;
-  }
-  __host__ __device__ volatile __half& operator=(const __half_raw& x) volatile {
-    data = x.data;
-    return *this;
-  }
-  volatile __half& operator=(const volatile __half_raw& x) volatile {
-    data = x.data;
-    return *this;
-  }
-  __half& operator=(__half_raw&& x) {
-    data = x.data;
-    return *this;
-  }
-  volatile __half& operator=(__half_raw&& x) volatile {
-    data = x.data;
-    return *this;
-  }
-  volatile __half& operator=(volatile __half_raw&& x) volatile {
-    data = x.data;
-    return *this;
-  }
-#if !defined(__HIP_NO_HALF_CONVERSIONS__)
-  template <typename T, Enable_if_t<std::is_floating_point<T>{}>* = nullptr>
-  __host__ __device__ __half& operator=(T x) {
-    data = static_cast<_hip_f16>(x);
-    return *this;
-  }
-#endif
+#define __ocml_exp10_f16(__X) exp10_h(__X)
+#define __ocml_exp10_2f16(__X) exp10_2h(__X)
 
-// MANIPULATORS - DEVICE ONLY
-#if !defined(__HIP_NO_HALF_CONVERSIONS__)
-  template <typename T, Enable_if_t<std::is_integral<T>{}>* = nullptr>
-  __device__ __half& operator=(T x) {
-    data = static_cast<_hip_f16>(x);
-    return *this;
-  }
-#endif
+#define __ocml_log2_f16(__X) log2_h(__X)
+#define __ocml_log2_2f16(__X) log2_2h(__X)
 
-#if !defined(__HIP_NO_HALF_OPERATORS__)
-  __device__ __half& operator+=(const __half& x) {
-    data += x.data;
-    return *this;
-  }
-  __device__ __half& operator-=(const __half& x) {
-    data -= x.data;
-    return *this;
-  }
-  __device__ __half& operator*=(const __half& x) {
-    data *= x.data;
-    return *this;
-  }
-  __device__ __half& operator/=(const __half& x) {
-    data /= x.data;
-    return *this;
-  }
-  __device__ __half& operator++() {
-    ++data;
-    return *this;
-  }
-  __device__ __half operator++(int) {
-    __half tmp{*this};
-    ++*this;
-    return tmp;
-  }
-  __device__ __half& operator--() {
-    --data;
-    return *this;
-  }
-  __device__ __half operator--(int) {
-    __half tmp{*this};
-    --*this;
-    return tmp;
-  }
-#endif
+#define __ocml_log_f16(__X) log_h(__X)
+#define __ocml_log_2f16(__X) log_2h(__X)
 
-// ACCESSORS
-#if !defined(__HIP_NO_HALF_CONVERSIONS__)
-  template <typename T, Enable_if_t<std::is_floating_point<T>{} &&
-                                    !std::is_same<T, double>{}>* = nullptr>
-  operator T() const {
-    return data;
-  }
-#endif
-  __host__ __device__ operator __half_raw() const { return __half_raw{data}; }
-  __host__ __device__ operator volatile __half_raw() const volatile {
-    return __half_raw{data};
-  }
-
-// ACCESSORS - DEVICE ONLY
-#if !defined(__HIP_NO_HALF_CONVERSIONS__)
-  template <typename T, Enable_if_t<std::is_integral<T>{}>* = nullptr>
-  __device__ operator T() const {
-    return data;
-  }
+#define __ocml_log10_f16(__X) log10_h(__X)
+#define __ocml_log10_2f16(__X) log10_2h(__X)
```

Comment:
What is `log10_h`? I can't find anything by this name besides this reference. 
The closest thing I see is a placeholder in `bfloat16_math.hh` 
```
// __device__​ __nv_bfloat16 	hlog10 ( const __nv_bfloat16 a )
```

is `bfloat16` and `fp16` two different things?

Reply:
To answer my own question, yes: 

https://nhigham.com/2018/12/03/half-precision-arithmetic-fp16-versus-bfloat16/

Reply:
Likely we haven't implemented it. Please bear in mind that this is a direct copy from AMDHIP and we better keep it so (make minimal unnecessary modifications to it, including formatting) in order to be able to easily synch updates from upstream later.

Reply:
Ok. It doesn't seem right that we need to be copying stuff over from hipamd. I thought the whole idea of splitting HIP-COMMON and HIP-AMD was to avoid stuff like this.

Reply:
I definitely agree this should be COMMON material when going forward.  I have another similar one in thepipeline, bf16.

Reply:
Somewhat relevant answer from Rahul as to why the math headers aren't shared:

```
It was decided to include these device headers as part of each backend for example -[ https://github.com/ROCm-Developer-Tools/hipamd/blob/develop/include/hip/amd_detail/amd_device_functions.h](https://github.com/ROCm-Developer-Tools/hipamd/blob/develop/include/hip/amd_detail/amd_device_functions.h). One of the reason was usage of inline functions.
```

Reply:
Why inline functions prevent putting to common?

Reply:
I don't know.